File: j-LIB-core/src/main/java/io/github/dot166/jlib/utils/VersionUtils.java
package io.github.dot166.jlib.utils;

import android.content.Context;
import android.os.Build;

import androidx.annotation.ChecksSdkIntAtLeast;

import io.github.dot166.jlib.R;

public class VersionUtils {

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.O_MR1)
    public static boolean isAtLeastO_1() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.P)
    public static boolean isAtLeastP() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.P;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.Q)
    public static boolean isAtLeastQ() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.R)
    public static boolean isAtLeastR() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.R;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.S)
    public static boolean isAtLeastS() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.S;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.S_V2)
    public static boolean isAtLeastS_1() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.S_V2;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.TIRAMISU)
    public static boolean isAtLeastT() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
    public static boolean isAtLeastU() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE;
    }

    @ChecksSdkIntAtLeast(api = Build.VERSION_CODES.VANILLA_ICE_CREAM)
    public static boolean isAtLeastV() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.VANILLA_ICE_CREAM;
    }

    // TODO: add check for Android Baklava once it has released
    // TODO: use the new SDK_INT_FULL variable for sdk checks once Android Baklava releases

    public static String getAndroidVersion() {
        if (isAtLeastT()) {
            return Build.VERSION.RELEASE_OR_PREVIEW_DISPLAY;
        }
        return Build.VERSION.RELEASE;
    }

    public static String getLibVersion(Context context) {
        return context.getString(R.string.lib_ver);
    }
}

File: j-LIB-core/src/main/java/io/github/dot166/jlib/utils/VersionUtils.kt
package io.github.dot166.jlib.utils

import android.os.Build
import androidx.annotation.ChecksSdkIntAtLeast
import io.github.dot166.jlib.version

@Deprecated("VersionUtils is no longer used")
object VersionUtils {

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to check this")
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES_FULL.R)
    val isAtLeastR: Boolean
        get() = Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.R

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to check this")
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES_FULL.S)
    val isAtLeastS: Boolean
        get() = Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.S

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to check this")
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES_FULL.S_V2)
    val isAtLeastS_1: Boolean
        get() = Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.S_V2

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to check this")
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES_FULL.TIRAMISU)
    val isAtLeastT: Boolean
        get() = Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.TIRAMISU

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to check this")
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES_FULL.UPSIDE_DOWN_CAKE)
    val isAtLeastU: Boolean
        get() = Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.UPSIDE_DOWN_CAKE

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to check this")
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES_FULL.VANILLA_ICE_CREAM)
    val isAtLeastV: Boolean
        get() = Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.VANILLA_ICE_CREAM

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to check this")
    @get:ChecksSdkIntAtLeast(api = Build.VERSION_CODES_FULL.BAKLAVA)
    val isAtLeastBaklava: Boolean
        get() = Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.BAKLAVA

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use the AndroidSDK to get the android version")
    val androidVersion: String?
        get() {
            if (isAtLeastT) {
                return Build.VERSION.RELEASE_OR_PREVIEW_DISPLAY
            }
            return Build.VERSION.RELEASE
        }

    @JvmStatic
    @Deprecated("VersionUtils is no longer used, please use io.github.dot166.jlib.version for jLib version")
    val libVersion: String
        get() = version
}

